lib/statsd/instrument/batched_udp_sink.rb

```diff
@@ -0,0 +1,118 @@
+      DEFAULT_FLUSH_INTERVAL = 1.0
+      MAX_PACKET_SIZE = 508
+
+      def self.for_addr(addr, flush_interval: DEFAULT_FLUSH_INTERVAL)
+        host, port_as_string = addr.split(":", 2)
+        new(host, Integer(port_as_string), flush_interval: flush_interval)
+      end
+
+      attr_reader :host, :port
+
+      def initialize(host, port, flush_interval: DEFAULT_FLUSH_INTERVAL)
+        @host = host
+        @port = port
+        @mutex = Mutex.new
+        @socket = nil
+        @flush_interval = flush_interval
+
+        require "concurrent/array"
+        @buffer = Concurrent::Array.new
+        @dispatcher_thread = nil
+        spawn_dispatcher
+      end
+
+      def sample?(sample_rate)
+        sample_rate == 1 || rand < sample_rate
+      end
+
+      def <<(datagram)
+        @buffer << datagram
+        spawn_dispatcher unless @dispatcher_thread&.alive?
+        self
+      end
+
+      def shutdown
+        @dispatcher_thread&.kill
+      end
+
+      private
+
+      def spawn_dispatcher
+        @dispatcher_thread = Thread.new { dispatch }
+      end
+
+      NEWLINE = "\n".b.freeze
+      def flush
+        return if @buffer.empty?
+
+        datagrams = @buffer.shift(@buffer.size)
+
+        until datagrams.empty?
+          packet = String.new(datagrams.pop, encoding: Encoding::BINARY, capacity: MAX_PACKET_SIZE)
+
+          until datagrams.empty? || packet.bytesize + datagrams.first.bytesize + 1 > MAX_PACKET_SIZE
+            packet << NEWLINE << datagrams.shift
+          end
+
+          send_packet(packet)
+        end
+      end
+
+      def dispatch
+        loop do
+          begin
+            start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+            flush
+            next_sleep_duration = @flush_interval - (Process.clock_gettime(Process::CLOCK_MONOTONIC) - start)
+            sleep(next_sleep_duration) if next_sleep_duration > 0
+          rescue => error
+            report_error(error)
+          end
+        end
+      end
+
+      def report_error(error)
+        StatsD.logger.error do
+          "[#{self.class.name}] The dispatcher thread encountered an error #{error.class}: #{error.message}"
+        end
+      end
+
+      def send_packet(packet)
+        retried = false
+        socket.send(packet, 0)
+      rescue SocketError, IOError, SystemCallError => error
+        StatsD.logger.debug do
+          "[#{self.class.name}] Resseting connection because of #{error.class}: #{error.message}"
+        end
+        invalidate_socket
+        if retried
+          StatsD.logger.warning do
+            "[#{self.class.name}] Events were dropped because of #{error.class}: #{error.message}"
+          end
+        else
+          retried = true
+          retry
+        end
+      end
+
+      def socket
+        @socket ||= begin
+          socket = UDPSocket.new
+          socket.connect(@host, @port)
+          socket
+        end
+      end
+
+      def invalidate_socket
+        @socket = nil
```

Comment:
If for some reason the app is creating short lived sinks, then just relying on garbage collection to close the sockets could result in file descriptors to be exhausted in between garbage collection cycles.  Did you avoid closing the socket to avoid getting an error with sending to the socket from the thread after it reads from this instance variable?

Reply:
I don't think short-lived periodically flushed sinks make sense outside of tests, so we could probably ignore that case for now.

Inside tests, we use `shutdown` to cleanup sinks, but we don't close the socket from that method.

We also call this method internally from `send_packet` on an error, which is a case where we should be closing the socket.